File: path-sum-ii.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > ret;
    void get_sum(TreeNode *root,int sum,vector<int>& v){
        if(root){
            v.push_back(root->val);
            if(root->val == sum && root->left == NULL && root->right == NULL){
                ret.push_back(v);
                v.pop_back();
                return ;
            }
            if(root->left)
                get_sum(root->left,sum-root->val,v);
            if(root->right)
                get_sum(root->right,sum - root->val,v);
            v.pop_back();
        }
    }
    vector<vector<int> > pathSum(TreeNode *root, int sum) {
        vector<int> v;
        get_sum(root,sum,v);
        return ret;
    }
};

File: python/tree/path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def dfs(self,root,sum,cur_arr):
        if root.left is None and root.right is None:
            if root.val == sum:
                self.result.append(cur_arr)
        else:
            if root.left:
                self.dfs(root.left, sum - root.val,cur_arr + [root.left.val])
            if root.right:
                self.dfs(root.right, sum - root.val,cur_arr + [root.right.val])


    def pathSum(self, root, sum):
        self.result = []

        if root:
            self.dfs(root,sum,[root.val])

        return self.result
